rust/src/transformations/make_stable_expr/expr_sum/mod.rs

```diff
@@ -0,0 +1,347 @@
+        DataType::Int16 => stability_map::<MI, P, i16>(series_domain, input_margin),
+        DataType::Int32 => stability_map::<MI, P, i32>(series_domain, input_margin),
+        DataType::Int64 => stability_map::<MI, P, i64>(series_domain, input_margin),
+        DataType::Float32 => stability_map::<MI, P, f32>(series_domain, input_margin),
+        DataType::Float64 => stability_map::<MI, P, f64>(series_domain, input_margin),
+        _ => fallible!(MakeTransformation, "unsupported data type"),
+    }?;
+
+    t_prior
+        >> Transformation::new(
+            input_domain,
+            output_domain,
+            Function::new_expr(Expr::sum),
+            input_metric.clone(),
+            LpDistance::default(),
+            stability_map,
+        )?
+}
+
+trait SumStabilityMap<MI: Metric, MO: Metric> {
+    fn make_stability_map(
+        margin: Margin,
+        input_metric: MI,
+        output_metric: MO,
+        max_size: usize,
+    ) -> Fallible<StabilityMap<MI, MO>>
+    where
+        (ExprDomain, MI): MetricSpace,
+        (ExprDomain, MO): MetricSpace;
+}
+
+// 1. spread into triple
+// aggregate  | l0    l1    li
+// select     | 1     d_in  d_in
+
+// 2. define map(d)
+// known N    | d // 2 * (U - L)
+// unknown N  | d * max(|L|, U)
+
+// 3. ideal sensitivity
+// l1distance | min(map(l1), l0 * map(li))
+// l2distance | min(map(l1), sqrt(l0) * map(li))
+
+// 4. max changed partitions
+//             SymmetricDistance     InsertDeleteDistance
+// aggregate  | max_partitions      | max_partitions.min(l0)
+// select     | 1                   | 1
+
+// 5. relaxation
+// l1distance | mcp       * relaxation(max_partition_len, L, U)
+// l2distance | sqrt(mcp) * relaxation(max_partition_len, L, U)
+
+// 6. actual sensitivity = ideal sensitivity + relaxation
+
+trait SumOutputMetric: Metric<Distance = Scalar> {}
+impl SumOutputMetric for L1Distance<Scalar> {}
+impl SumOutputMetric for L2Distance<Scalar> {}
+
+// 3. ideal sensitivity
+trait NativeSumOutputMetric: Metric {
+    type SumOutputMetric: SumOutputMetric;
+    fn ideal_sensitivity(
+        l0: Self::Distance,
+        l1: Self::Distance,
+        li: Self::Distance,
+    ) -> Fallible<Self::Distance>;
+}
+
+impl<Q: Number> NativeSumOutputMetric for L1Distance<Q> {
+    type SumOutputMetric = L1Distance<Scalar>;
+    fn ideal_sensitivity(
+        l0: Self::Distance,
+        l1: Self::Distance,
+        li: Self::Distance,
+    ) -> Fallible<Self::Distance> {
+        l1.total_min(l0.inf_mul(&li)?)
+    }
+}
+
+impl<Q: Float> NativeSumOutputMetric for L2Distance<Q> {
+    type SumOutputMetric = L2Distance<Scalar>;
+    fn ideal_sensitivity(
+        l0: Self::Distance,
+        l1: Self::Distance,
+        li: Self::Distance,
+    ) -> Fallible<Self::Distance> {
+        l1.total_min(l0.inf_sqrt()?.inf_mul(&li)?)
+    }
+}
+
+pub trait Summand: NumericDataType + CheckAtom + ProductOrd {
+    type Sum: Accumulator + InfCast<Self> + ProductOrd;
+}
+
+// ```python
```

Comment:
More explanation: Is this pseudo code that the rust implements? Is this a demonstration of what the python API will look like?

Reply:
Added comments.